test/benchmark_http_client_test.cc

```diff
@@ -46,41 +67,46 @@ class BenchmarkClientHttpTest : public Test {
 
     auto& tracer = static_cast<Envoy::Tracing::MockHttpTracer&>(*http_tracer_);
     EXPECT_CALL(tracer, startSpan_(_, _, _, _))
-        .WillRepeatedly(
-            Invoke([&](const Envoy::Tracing::Config& config, const Envoy::Http::HeaderMap&,
-                       const Envoy::StreamInfo::StreamInfo&,
-                       const Envoy::Tracing::Decision) -> Envoy::Tracing::Span* {
-              EXPECT_EQ(Envoy::Tracing::OperationName::Egress, config.operationName());
-              auto* span = new NiceMock<Envoy::Tracing::MockSpan>();
-              return span;
-            }));
-    request_generator_ = []() {
-      auto header = std::make_shared<Envoy::Http::TestRequestHeaderMapImpl>(
-          std::initializer_list<std::pair<std::string, std::string>>(
-              {{":scheme", "http"}, {":method", "GET"}, {":path", "/"}, {":host", "localhost"}}));
-      return std::make_unique<RequestImpl>(header);
-    };
+        .WillRepeatedly([](const Envoy::Tracing::Config& config, const Envoy::Http::HeaderMap&,
+                           const Envoy::StreamInfo::StreamInfo&,
+                           const Envoy::Tracing::Decision) -> Envoy::Tracing::Span* {
+          EXPECT_EQ(Envoy::Tracing::OperationName::Egress, config.operationName());
+          auto* span = new NiceMock<Envoy::Tracing::MockSpan>();
+          return span;
+        });
   }
 
-  void testBasicFunctionality(const uint64_t max_pending, const uint64_t connection_limit,
-                              const uint64_t amount_of_request) {
+  // Primary testing method. Confirms that connection limits are met and number of requests are
+  // correct. If header expectations is not null, also checks the header expectations, if null, it
+  // is ignored.
+  void TestBenchmarkClientProcessesExpectedInflightRequests(
+      const uint64_t max_pending, const uint64_t connection_limit, const uint64_t amount_of_request,
+      const RequestGenerator& request_generator,
+      const absl::flat_hash_set<std::string>* header_expectations = nullptr) {
```

Comment:
Optional as far as I'm concerned, but we could pass a reference instead of a pointer here by creating an overload which doesn't have the expectations, and then calls this with an empty set. Pointers are mostly avoided, and doing this would help consistency.

Reply:
I originally had the empty set thing, but I felt like nullable was clearer, since empty set might mislead people to think that we're testing that we get an empty set of headers, rather than that we don't test it.

Reply:
Fair enough

Reply:
Seeing a const pointer as an argument is fairly unusual. If the intent is to make this an optional argument, we should use `absl::optional` instead and the value can then be the expected const reference (or non-const reference). Otherwise the reader of this code has to guess why we are using a pointer here.

There are already a few places in the Nighthawk codebase that use `absl::optional` and can serve as an example.